solidity/random-beacon/README.adoc

```diff
@@ -1 +1,324 @@
+start the off-chain protocol to sign the previous relay entry producing a new one.
+
+Off-chain clients are expected to follow the order when submitting relay entry
+to minimize and distribute costs evenly, as described in
+<<fees-and-rewards,Fees and Rewards>> section but no ordering is enforced
+on-chain. New relay entry should be submitted using 
+`RandomBeaocn.submitRelayEntry(bytes calldata entry)`
+function.
+
+=== Callbacks
+
+Random Beacon supports simple, low gas budget callbacks from a relay entry
+submit a transaction with a gas limit being a governable parameter.
+
+
+When requesting a relay entry, it is possible to pass an optional address
+parameter - this is the address of a contract implementing
+`IRandomBeaconConsumer` interface that should be called when a new relay entry
+is submitted to the chain.
+
+Smart contract consuming new relay entry needs to implement `IRandomBeaconConsumer`
+interface. The gas limit for `__beaconCallback` is initially set to 50k gas
+which is enough to `SSTORE` new relay entry, block height in which the entry was
+submitted, and to emit an event. Callback gas limit is a governable value.
+Failure in the callback function does not revert the relay entry transaction.
+
+```solidity
+interface IRandomBeaconConsumer {
+    /// @notice Receives relay entry produced by Keep Random Beacon. This function
+    /// should be called only by Keep Random Beacon.
+    ///
+    /// @param relayEntry Relay entry (random number) produced by Keep Random
+    ///                   Beacon.
+    /// @param blockNumber Block number at which the relay entry was submitted
+    ///                    to the chain.
+    function __beaconCallback(uint256 relayEntry, uint256 blockNumber) external;
+}
+```
+
+=== Timeouts
+
+There are two timeouts for a relay entry to be provided by a group: soft timeout
+and hard timeout.
+
+==== Soft Relay Entry Timeout
+
+The soft timeout is the group size multiplied by the number of blocks for a
+member to become eligible to submit a relay entry. Eligibility is not enforced
+on-chain but off-chain clients are expected to agree and follow it.
+
+If no entry was provided within the soft timeout, all operators in the group
+start bleeding and losing their stake. The bleeding increases linearly from 0 to
+the governable slashing amount per operator over time, until the hard timeout is
+reached or until a relay entry is submitted by the group.
+
+The time for a single group member to become eligible to submit a result and the
+hard relay entry timeout are governable parameters. This gives a chance to start
+with more forgiving penalties and increase them over time. In general, the
+slashing penalty should be proportional to rewards and the frequency of relay
+requests and associated risk.
+
+==== Hard Relay Entry Timeout
+
+When the hard timeout is reached, anyone can notify about this fact by calling
+`RandomBeacon.reportRelayEntryTimeout()` function and receiving a notifier
```

Comment:
anyone can notify ... and receiving  -> anyone can notify ... and receive

Reply:
cf87b67e3d6ef59dfd1b5263c1f62d3b0204b873